lib/internal/modules/esm/translators.js

```diff
@@ -125,23 +137,106 @@ translators.set('module', async function moduleStrategy(url) {
 // Strategy for loading a node-style CommonJS module
 const isWindows = process.platform === 'win32';
 const winSepRegEx = /\//g;
-translators.set('commonjs', function commonjsStrategy(url, isMain) {
+translators.set('commonjs', async function commonjsStrategy(url, isMain) {
   debug(`Translating CJSModule ${url}`);
-  return new ModuleWrap(url, undefined, ['default'], function() {
+
+  let filename = internalURLModule.fileURLToPath(new URL(url));
+  if (isWindows)
+    filename = StringPrototypeReplace(filename, winSepRegEx, '\\');
+
+  // Apply CJS exports parser.
+  if (!cjsParserInitialized) {
+    await cjsInit();
+    cjsParserInitialized = true;
+  }
+  const { module, exportNames } = cjsPreparseModuleExports(filename);
+  const namesWithDefault = exportNames.has('default') ?
+    [...exportNames] : ['default', ...exportNames];
+
+  return new ModuleWrap(url, undefined, namesWithDefault, function() {
     debug(`Loading CJSModule ${url}`);
-    const pathname = internalURLModule.fileURLToPath(new URL(url));
+
     let exports;
-    const cachedModule = CJSModule._cache[pathname];
-    if (cachedModule && asyncESM.ESMLoader.cjsCache.has(cachedModule)) {
-      exports = asyncESM.ESMLoader.cjsCache.get(cachedModule);
-      asyncESM.ESMLoader.cjsCache.delete(cachedModule);
+    if (asyncESM.ESMLoader.cjsCache.has(module)) {
+      exports = asyncESM.ESMLoader.cjsCache.get(module);
+      asyncESM.ESMLoader.cjsCache.delete(module);
     } else {
-      exports = CJSModule._load(pathname, undefined, isMain);
+      exports = CJSModule._load(filename, undefined, isMain);
+    }
+
+    // Verify that it really does set __esModule, not just statically.
+    const esModule = ObjectPrototypeHasOwnProperty(exports, '__esModule') &&
+        exports.__esModule;
+    if (esModule) {
+      for (const exportName of exportNames) {
+        if (!ObjectPrototypeHasOwnProperty(exports, exportName) ||
+            exportName === 'default')
+          continue;
+        // We might trigger a getter -> dont fail.
+        let value;
+        try {
+          value = exports[exportName];
+        } catch {}
+        this.setExport(exportName, value);
+      }
     }
     this.setExport('default', exports);
   });
 });
 
+function cjsPreparseModuleExports(filename) {
+  let module = CJSModule._cache[filename];
+  if (module) {
+    const cached = cjsParseCache.get(module);
+    if (cached)
+      return { module, exportNames: cached.exportNames };
+  }
+  const loaded = Boolean(module);
+  if (!loaded) {
+    module = new CJSModule(filename);
+    module.filename = filename;
+    module.paths = CJSModule._nodeModulePaths(module.path);
+    CJSModule._cache[filename] = module;
+  }
+
+  let source;
+  try {
+    source = readFileSync(filename, 'utf8');
```

Comment:
I _think_ we can achieve somehow better perf if this was `async`.. could we do more than one of these in parallel while loading esm?

Reply:
There have been experiments in the past on async loading in the es module loader that found using async fs calls slowed down the startup time. @bmeck had some information on this before I believe.

All fs calls for all module loading in the ES module loader are sync for this reason, although that could be reconsidered if a new benchmark could show savings here.